### PR TITLE
Allow overriding of Objective-C message send P/Invokes

### PR DESCRIPTION
**Pull request by elinor-fung**
- Add `Bridge.SetMessageSendCallback`
- If message send functions are set, use P/Invoke override to redirect to specified function(s)

### CODE REVIEW

File: src/coreclr/vm/interoplibinterface_objc.cpp [hunk 1]
```diff
@@ -11,20 +11,68 @@
...
+#include <pinvokeoverride.h>
+
+#define OBJC_MSGSEND "objc_msgSend"
+
+namespace
+{
+    BOOL s_msgSendOverridden = FALSE;
+    void* s_msgSendOverrides[ObjCWrappersNative::MsgSendFunction::Last + 1] = {0};
+
+    const char* ObjectiveCLibrary = "/usr/lib/libobjc.dylib";
```
Comment (elinor-fung) [inline src/coreclr/vm/interoplibinterface_objc.cpp:23]: Took this from:
https://github.com/xamarin/xamarin-macios/blob/13fd8f5b07361a14186d77e2b1719449414edb2f/runtime/runtime.m#L2451-L2454

The [doc](https://docs.microsoft.com/xamarin/ios/internals/objective-c-selectors) examples use `Constants.ObjectiveCLibrary`:
https://github.com/xamarin/xamarin-macios/blob/13fd8f5b07361a14186d77e2b1719449414edb2f/src/ObjCRuntime/Constants.cs#L9

Do we care to do all the possible variations too? (e.g. `objc`, `objc.dylib`, `libobjc`, `libobjc.dylib`)

File: src/coreclr/vm/interoplibinterface_objc.cpp [hunk 2]
```diff
@@ -11,20 +11,68 @@
...
-    _In_ void* fptr_objc_msgSendSuper,
-    _In_ void* fptr_objc_msgSendSuper_stret)
+#include <pinvokeoverride.h>
+
+#define OBJC_MSGSEND "objc_msgSend"
+
+namespace
+{
+    BOOL s_msgSendOverridden = FALSE;
+    void* s_msgSendOverrides[ObjCWrappersNative::MsgSendFunction::Last + 1] = {0};
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/interoplibinterface_objc.cpp:21]: Technically `{}` is sufficient.

File: src/coreclr/vm/interoplibinterface_objc.cpp [hunk 3]
```diff
@@ -11,20 +11,68 @@
...
+    const char* MsgSendEntryPoints[ObjCWrappersNative::MsgSendFunction::Last + 1] =
+    {
+        OBJC_MSGSEND,
+        OBJC_MSGSEND "_fpret",
+        OBJC_MSGSEND "_stret",
+        OBJC_MSGSEND "Super",
+        OBJC_MSGSEND "Super_stret",
+    };
+
+    const void* MessageSendPInvokeOverride(const char* libraryName, const char* entrypointName)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/interoplibinterface_objc.cpp:33]: `void*`. Ca we get SAL on the args.

File: src/coreclr/vm/interoplibinterface_objc.cpp [hunk 4]
```diff
@@ -11,20 +11,68 @@
...
     QCALL_CONTRACT;
 
+    bool success;
+
     BEGIN_QCALL;
 
+    success = FastInterlockCompareExchangePointer(&s_msgSendOverrides[msgSendFunction], fptr, NULL) == NULL;
+
+    // Set override if we haven't already
+    if (success && FALSE == FastInterlockCompareExchange((LONG*)&s_msgSendOverridden, TRUE, FALSE))
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/interoplibinterface_objc.cpp:70]: This is setting the override callback right? Let's comment on that and validate the enum passed in is within range. That could be asserted here instead and the actual error case handled in managed if desired.

File: src/coreclr/vm/interoplibinterface_objc.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/interoplibinterface_objc.cpp:23]: I am fine with the strict supported scenario for this. It is a quirk behavior so we can be strict.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/ObjectiveC/Bridge.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        /// <remarks>
+        /// Providing an override can enable support for Objective-C
+        /// exception propagation and variadic argument support.
+        /// </remarks>
+        public static void SetMessageSendCallback(MsgSendFunction msgSendFunction, IntPtr func)
+        {
+            if (func == IntPtr.Zero)
+                throw new ArgumentNullException(nameof(func));
+
+            if (!Enum.IsDefined<MsgSendFunction>(msgSendFunction))
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/ObjectiveC/Bridge.cs:41]: From .NET Framework design guidelines (ebook available in ms library): "DO NOT use Enum.IsDefined for enum range checks."

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
So many "P/Inovkes"!

File: src/coreclr/vm/ecalllist.h [hunk 6]
```diff
@@ -1107,6 +1109,9 @@ FCClassElement("AssemblyExtensions", "System.Reflection.Metadata", gAssemblyExte
 FCClassElement("AssemblyLoadContext", "System.Runtime.Loader", gAssemblyLoadContextFuncs)
 
 FCClassElement("AssemblyName", "System.Reflection", gAssemblyNameFuncs)
+#ifdef FEATURE_OBJCWRAPPERS
```
Comment (jkoritzinsky) [inline src/coreclr/vm/ecalllist.h:1112]: Do we want to keep the name `FEATURE_OBJCWRAPPERS` now that we've renamed the API?

**Review by jkoritzinsky — APPROVED** [review]
LGTM

File: src/coreclr/vm/ecalllist.h (hunk 6 above)
Comment (elinor-fung) [reply src/coreclr/vm/ecalllist.h:1112]: `FEATURE_OBJCBRIDGE`?

Comment (elinor-fung) [reply src/coreclr/vm/ecalllist.h:1112]: I'll do a follow-up change of renaming the define and deleting the current `Wrappers` class (and friends).

Comment (jkoritzinsky) [reply src/coreclr/vm/ecalllist.h:1112]: I like that name.

**Pull request merged**